crates/ty_python_semantic/resources/mdtest/annotations/self.md

```diff
@@ -53,6 +48,144 @@ class Outer:
             return self
 ```
 
+## Type of (unannotated) `self` parameters
+
+In instance methods, the first parameter (regardless of its name) is assumed to have the type
+`typing.Self`, unless it has an explicit annotation. This does not apply to `@classmethod` and
+`@staticmethod`s.
+
+```toml
+[environment]
+python-version = "3.11"
+```
+
+```py
+from typing import Self
+
+class A:
+    def implicit_self(self) -> Self:
+        # TODO: This should be Self@implicit_self
+        reveal_type(self)  # revealed: Unknown
+
+        return self
+
+    def a_method(self) -> int:
+        def first_arg_is_not_self(a: int) -> int:
+            reveal_type(a)  # revealed: int
+            return a
+        return first_arg_is_not_self(1)
+
+    @classmethod
+    def a_classmethod(cls) -> Self:
+        # TODO: This should be type[Self@bar]
+        reveal_type(cls)  # revealed: Unknown
+        return cls()
+
+    @staticmethod
+    def a_staticmethod(x: int): ...
+
+a = A()
+
+reveal_type(a.implicit_self())  # revealed: A
+reveal_type(a.implicit_self)  # revealed: bound method A.implicit_self() -> A
+```
+
+Calling an instance method explicitly verifies the first argument:
+
+```py
+A.implicit_self(a)
+
+# error: [invalid-argument-type] "Argument to function `implicit_self` is incorrect: Argument type `Literal[1]` does not satisfy upper bound `A` of type variable `Self`"
+A.implicit_self(1)
+```
+
+If the method is a class or static method then first argument is not self:
+
+```py
+A.a_classmethod()
+
+A.a_classmethod(a)  # error: [too-many-positional-arguments]
+
+A.a_staticmethod(1)
+a.a_staticmethod(1)
+
+A.a_staticmethod(a)  # error: [invalid-argument-type]
+```
+
+The first parameter of instance methods always has type `Self`, if it is not explicitly annotated.
+The name `self` is not special in any way.
+
+```py
+class B:
+    def name_does_not_matter(this) -> Self:
+        # TODO: Should reveal Self@name_does_not_matter
+        reveal_type(this)  # revealed: Unknown
+
+        return this
+
+    def positional_only(self, /, x: int) -> Self:
+        # TODO: Should reveal Self@positional_only
+        reveal_type(self)  # revealed: Unknown
+        return self
+
+    def keyword_only(self, *, x: int) -> Self:
+        # TODO: Should reveal Self@keyword_only
+        reveal_type(self)  # revealed: Unknown
+        return self
+
+    @property
+    def a_property(self) -> Self:
+        # TODO: Should reveal Self@a_property
+        reveal_type(self)  # revealed: Unknown
+        return self
+
+reveal_type(B().name_does_not_matter())  # revealed: B
+
+# TODO: this should be B
+reveal_type(B().positional_only(1))  # revealed: Unknown
+
+reveal_type(B().keyword_only(x=1))  # revealed: B
+
+# TODO: this should be B
+reveal_type(B().a_property)  # revealed: Unknown
```

Comment:
This should also be fixed. I have not looked into it yet. Maybe it can be deferred to a follow-up.